### PR TITLE
[ALLUXIO-3147] Enable more commands to use wildcards.

### PR DESCRIPTION
**Pull request by yuzhu**
https://alluxio.atlassian.net/browse/ALLUXIO-3147

Refactored wildcard command to enable more commands to use wildcard in their URIs.
All commands are now children of AbstractFileSystemCommand.

### CODE REVIEW

**Comment by yuzhu** [comment]
alluxio-bot, check this please

File: shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java [hunk 1]
```diff
@@ -28,4 +37,29 @@
...
+
+  protected void runPath(AlluxioURI plainPath) throws AlluxioException, IOException{
+  }
+
+  protected void runWildCardCmd(AlluxioURI wildCardPath) throws IOException {
+    List<AlluxioURI> paths = FileSystemShellUtils.getAlluxioURIs(mFileSystem, wildCardPath);
+    if (paths.size() == 0) { // A unified sanity check on the paths
+      throw new IOException(wildCardPath + " does not exist.");
+    }
+    Collections.sort(paths, FileSystemShellUtils.createAlluxioURIComparator());
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java]: try this: `paths.sort(Comparator.comparing(AlluxioURI::getPath))`

File: shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java [hunk 2]
```diff
@@ -40,6 +41,11 @@ public CopyFromLocalCommand(FileSystem fs) {
     mCpCommand = new CpCommand(fs);
   }
 
+  @Override
+  protected void runPath(AlluxioURI plainURI) throws AlluxioException, IOException {
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java]: do we need this?

File: shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java [hunk 3]
```diff
@@ -39,6 +40,11 @@ public CopyToLocalCommand(FileSystem fs) {
     mCpCommand = new CpCommand(fs);
   }
 
+  @Override
+  protected void runPath(AlluxioURI plainURI) throws AlluxioException, IOException {
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java]: do we need this?

File: shell/src/main/java/alluxio/cli/fs/command/HeadCommand.java [hunk 4]
```diff
@@ -36,12 +36,14 @@
...
  */
 @ThreadSafe
-public final class HeadCommand extends WithWildCardPathCommand {
+public final class HeadCommand extends AbstractFileSystemCommand {
   private static final Option BYTES_OPTION = Option.builder("c")
-      .required(false)
-      .numberOfArgs(1)
-      .desc("number of bytes (e.g., 1024, 4KB)")
-      .build();
+          .required(false)
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/HeadCommand.java]: indent is off

File: shell/src/main/java/alluxio/cli/fs/command/LocationCommand.java [hunk 5]
```diff
@@ -47,10 +47,10 @@ public String getCommandName() {
   }
 
   @Override
-  protected void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(path);
+  protected void runPath(AlluxioURI plainPath) throws AlluxioException, IOException {
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/LocationCommand.java]: actually how about calling this method `runPlainPath`?

File: shell/src/main/java/alluxio/cli/fs/command/RmCommand.java [hunk 6]
```diff
@@ -78,9 +80,9 @@ public Options getOptions() {
   }
 
   @Override
-  protected void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioException, IOException {
+  protected void runPath(AlluxioURI path) throws AlluxioException, IOException {
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/RmCommand.java]: I think most classes do not passing of the commandLine object. So perhaps it is ok to design the interface to address the common case ?

File: shell/src/main/java/alluxio/cli/fs/command/TailCommand.java [hunk 7]
```diff
@@ -36,12 +36,14 @@
...
  */
 @ThreadSafe
-public final class TailCommand extends WithWildCardPathCommand {
+public final class TailCommand extends AbstractFileSystemCommand {
   private static final Option BYTES_OPTION = Option.builder("c")
-      .required(false)
-      .numberOfArgs(1)
-      .desc("number of bytes (e.g., 1024, 4KB)")
-      .build();
+          .required(false)
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/TailCommand.java]: fixed

File: shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java (hunk 1 above)
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java]: done

File: shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java (hunk 2 above)
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java]: deleted

File: shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java (hunk 3 above)
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java]: deleted

File: shell/src/main/java/alluxio/cli/fs/command/HeadCommand.java (hunk 4 above)
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/HeadCommand.java]: fixed

File: shell/src/main/java/alluxio/cli/fs/command/LocationCommand.java (hunk 5 above)
Comment (yuzhu) [reply shell/src/main/java/alluxio/cli/fs/command/LocationCommand.java]: done

**Comment by yuzhu** [comment]
The reason I deprecated WithWildCardPathCommand.java instead of deleted it, is because there is a deprecated class that depends on WithWildCardPathCommand.java

File: shell/src/main/java/alluxio/cli/fs/FileSystemShellUtils.java [hunk 8]
```diff
@@ -31,6 +31,7 @@
 import java.io.File;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Comparator;
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/FileSystemShellUtils.java]: remove?

File: shell/src/main/java/alluxio/cli/fs/FileSystemShellUtils.java [hunk 9]
```diff
@@ -282,4 +283,5 @@ private static boolean match(AlluxioURI fileURI, AlluxioURI patternURI) {
   protected static boolean match(String filePath, String patternPath) {
     return match(new AlluxioURI(filePath), new AlluxioURI(patternPath));
   }
+
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/FileSystemShellUtils.java]: ditto

File: shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java [hunk 10]
```diff
@@ -28,4 +37,29 @@
   protected AbstractFileSystemCommand(FileSystem fs) {
     mFileSystem = fs;
   }
+
+  protected void runPlainPath(AlluxioURI plainPath) throws AlluxioException, IOException{
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java]: please javadoc this method

File: shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java [hunk 11]
```diff
@@ -28,4 +37,29 @@
...
+  }
+
+  protected void runWildCardCmd(AlluxioURI wildCardPath) throws IOException {
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/AbstractFileSystemCommand.java]: please javadoc this method

File: shell/src/main/java/alluxio/cli/fs/command/ChecksumCommand.java [hunk 12]
```diff
@@ -43,6 +43,18 @@ public ChecksumCommand(FileSystem fs) {
     super(fs);
   }
 
+  @Override
+  protected void runPlainPath(AlluxioURI plainPath) throws AlluxioException, IOException {
+    URIStatus status = mFileSystem.getStatus(plainPath);
+    if (status.isFolder()) {
+      throw new FileDoesNotExistException(
+              ExceptionMessage.PATH_MUST_BE_FILE.getMessage(plainPath.getPath()));
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/ChecksumCommand.java]: 4-space indent please for line wrap

File: shell/src/main/java/alluxio/cli/fs/command/RmCommand.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (apc999) [reply shell/src/main/java/alluxio/cli/fs/command/RmCommand.java]: I looked through the code. I think actually passing `cli` as an arg in `runPlainPath` will simplify quite some implementations as we could remove these member variables we defined. Keep code simpler and cleaner is always helpful :)

File: shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java [hunk 13]
```diff
@@ -11,6 +11,7 @@
 
 package alluxio.cli.fs.command;
 
+import alluxio.AlluxioURI;
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/CopyFromLocalCommand.java]: ??

File: shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java [hunk 14]
```diff
@@ -11,6 +11,7 @@
 
 package alluxio.cli.fs.command;
 
+import alluxio.AlluxioURI;
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/CopyToLocalCommand.java]: remove

**Comment by yuzhu** [comment]
@apc999 @aaudiber could you please take a look at the latest version? thanks.

File: shell/src/main/java/alluxio/cli/fs/command/FreeCommand.java [hunk 15]
```diff
@@ -60,8 +61,9 @@ public String getCommandName() {
   }
 
   @Override
-  protected void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioException, IOException {
-    FreeOptions options = FreeOptions.defaults().setRecursive(true).setForced(cl.hasOption("f"));
+  protected void runPlainPath(AlluxioURI path, CommandLine cl)
+      throws AlluxioException, IOException {
+    FreeOptions options = FreeOptions.defaults().setRecursive(true).setForced(mIsForced);
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/FreeCommand.java]: `cl.hasOption("f")` to replace and remove `mIsForced`

File: shell/src/main/java/alluxio/cli/fs/command/LsCommand.java [hunk 16]
```diff
@@ -258,12 +258,23 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
   }
 
   @Override
-  public void runCommand(AlluxioURI path, CommandLine cl) throws AlluxioException, IOException {
+  protected void runPlainPath(AlluxioURI path, CommandLine cl)
+
+      throws AlluxioException, IOException {
     ls(path, cl.hasOption("R"), cl.hasOption("f"), cl.hasOption("d"), cl.hasOption("h"),
-        cl.hasOption("p"), cl.getOptionValue("sort", "name"),
+            cl.hasOption("p"), cl.getOptionValue("sort", "name"),
```
Comment (apc999) [inline shell/src/main/java/alluxio/cli/fs/command/LsCommand.java]: indent

**Pull request merged**